service-registry/src/main/java/org/apache/servicecomb/serviceregistry/task/ServiceCenterTask.java

```diff
@@ -62,7 +75,27 @@ public void onMicroserviceInstanceHeartbeatTask(MicroserviceInstanceHeartbeatTas
     if (task.getHeartbeatResult() != HeartbeatResult.SUCCESS) {
       LOGGER.info("read MicroserviceInstanceHeartbeatTask status is {}", task.taskStatus);
       onException();
+      if (!safeMode && consecutiveFailedTimes.incrementAndGet() > checkTimes) {
+        LOGGER.warn("service center is unavailable, enter safe mode");
+        eventBus.post(new SafeModeChangeEvent(true));
+      }
+      if (consecutiveSucceededTimes.get() != 0) {
+        consecutiveSucceededTimes.set(0);
+      }
+      return;
+    }
+    if (safeMode && consecutiveSucceededTimes.incrementAndGet() > checkTimes) {
+      LOGGER.warn("service center is recovery, exit safe mode");
+      eventBus.post(new SafeModeChangeEvent(false));
     }
+    if (consecutiveFailedTimes.get() != 0) {
+      consecutiveFailedTimes.set(0);
+    }
+  }
+
+  @Subscribe
+  public void onSafeModeChanged(SafeModeChangeEvent modeChangeEvent) {
```

Comment:
It's simpler to set safeMode=true before sending event, no need to subscript event here.

Reply:
done

Reply:
oh, yes. I agree with you